File: 030/SubstringwithConcatenationofAllWords.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SubstringwithConcatenationofAllWords {

    public List<Integer> findSubstring(String s, String[] words) {
        if (s == null || words == null) return new ArrayList<>();
        List<Integer> res = new ArrayList<>();
        int n = words.length;
        int m = words[0].length();
        HashMap<String, Integer> map = new HashMap<>();

        for (String str : words) {
            map.put(str, map.getOrDefault(str, 0)+1);
        }

        for (int i = 0; i <= s.length() - n * m; i++) {
            HashMap<String, Integer> copy = new HashMap<>(map);
            int k = n;
            int j = i;
            while (k > 0) {
                String str = s.substring(j, j+m);
                if (!copy.containsKey(str) || copy.get(str) < 1) {
                    break;
                }
                copy.put(str, copy.get(str)-1);
                k--;
                j += m;
            }
            if (k == 0) res.add(i);
        }
        return res;
    }

    public static void main(String[] args) {
        SubstringwithConcatenationofAllWords sol = new SubstringwithConcatenationofAllWords();
        String s = "barfoothefoobarman";
        String[] words = { "foo", "bar"};
        System.out.println(sol.findSubstring(s, words));
    }
}

File: 030/substring-with-concatenation-of-all-words.go
package main

import "fmt"

func findSubstring(s string, words []string) []int {
	sLen := len(s)
	wsLen := len(words)
	res := make([]int, 0, sLen)
	if sLen == 0 || wsLen == 0 {
		return res
	}

	wLen := len(words[0])
	if sLen < wsLen*wLen {
		return res
	}

	record := make(map[string]int, wsLen)
	for _, word := range words {
		if len(word) != wLen {
			return res
		}
		record[word]++
	}

	for i := 0; i <= sLen-wLen*wsLen; i++ {
		// make a copy of record
		m := make(map[string]int, wsLen)
		for k, v := range record {
			m[k] = v
		}

		k := wsLen
		j := i
		for k > 0 {
			str := s[j : j+wLen]
			count, ok := m[str]
			if !ok || count < 1 {
				break
			}
			m[str] = m[str] - 1
			k--
			j += wLen
		}
		if k == 0 {
			res = append(res, i)
		}
	}
	return res
}

func main() {
	//s := "barfoothefoobarman"
	//words := []string{"foo", "bar"}
	s := ""
	words := []string{}
	fmt.Println(findSubstring(s, words))
}
